lib/buffer.js

```diff
@@ -157,6 +172,26 @@ function showFlaggedDeprecation() {
   bufferWarningAlreadyEmitted = true;
 }
 
+// Calling this method does not affect existing buffers, only new ones.
+Buffer.harden = function() {
+  if (hardened) return;
+  if (isInsideNodeModules()) {
+    throw new ERR_ASSERTION(
+      'Buffer.harden() should be called only from the top-level module. ' +
+      'Calling Buffer.harden() from dependencies is not supported.'
+    );
+  }
+  hardened = true;
+  Object.defineProperty(Buffer, 'poolSize', {
+    enumerable: true,
+    get: () => 0,
+    set: () => {},
+  });
+  Object.freeze(Buffer);
+  Object.freeze(Buffer.prototype);
+  Object.freeze(module.exports);
```

Comment:
Frozen object are slower. How much slower in this case?

Reply:
I don't think that's true. It used to be a long time ago, but they should be the same as regular objects these days.

Reply:
Some indirect tests still shows regressions.

Reply:
@mcollina I will run benchmarks on this.

But I don't expect that frozen objects would be what would cause the most of performance hit here, `Buffer.harden()` also enables zero-filling for allocUnsafe and disables buffer pooling completely, I expect that to me more noticable.

This whole PR comes with an opt-in perf hit that would be documented once I get the numbers.
The primary targets for opt-in hardening are not performance-sensitive backend setups, but pretty much everything else.